MONGOID-4549 Correctly track dependents locally to model

saghm opened the pull request.

p-mongo commented:
Deleting a dependent college user with a dependent report card should delete the report card.

Also since mongoid allows this:

<pre>
class DoubleAssoc
  include Mongoid::Document

  has_many :deps, dependent: :destroy
end

class DoubleAssocOne < DoubleAssoc
  has_many :deps, dependent: :nullify, inverse_of: :double_assoc
end

class DoubleAssocTwo < DoubleAssocOne
  has_many :deps, dependent: :destroy, inverse_of: :double_assoc
end
</pre>

There should be a test that destroying DoubleAssocOne destroys its deps and destroying a DoubleAssocTwo doesn't (on master destroying DoubleAssocOne destroys its deps, if *this* is desired behavior there should be an exception raised when attempting to redefine dependent behavior).

saghm commented:
> on master destroying DoubleAssocOne destroys its deps, if this is desired behavior there should be an exception raised when attempting to redefine dependent behavior

This is another instance of the same bug. I'll add the two test cases that you suggest.

saghm commented:
Actually, from looking it over again, I don't think it's the same bug. I believe as implemented on master, dependents are cumulative, so there's both a `destroy` and `nullify` dependent for `deps` on `DoubleAssocOne`. I agree with you that `DoubleAssocOne` should not destroy its dependents and `DoubleAssocTwo` should, so I'll fix that if the test shows it doesn't work like that.

saghm commented:
@p-mongo do the additional tests look good to you?

p-mongo commented:
I am not seeing a test that deleting a dependent college user with a dependent report card deletes their report card, is it in the diff?

saghm commented:
As written before, CollegeUser doesn't even have a  reference to ReportCard, so I added the transitivity test for Student. Can you elaborate on why the test needs to be on CollegeUser?

p-mongo commented:
The reason why `class_attribute` exists is because simple instance variable assignment on class level, which is what this diff proposes, often times does not behave correctly in the context of inheritance. Hence, inheritance should be explicitly tested.

saghm commented:
In what way does CollegeUser use inheritance that Student doesn't? They both inherit from the same superclass. I'm really not understanding why you think that testing CollegeUser is better than testing Student when they're literally identical except that Student has the property (a dependent report card) that we actually want to test.

saghm commented:
Per offline discussion, I added in tests to make sure that the dependent behavior was defined properly across inheritance boundaries. You were correct that subclasses did not inherit the dependents from a superclass, so I added in a manual call to `superclass.dependents` in the initialization of the dependents array. To fix the duplicate dependents defined for a subclass that overrides a given dependent, I changed the `define_dependency!` method to remove existing dependents that are associated with the same class before adding the new one.

p-mongo commented on lib/mongoid/association/depending.rb:
```diff
@@ -10,7 +10,7 @@ module Depending
       included do
         def self.dependents
-          @dependents ||= []
+          @dependents ||= superclass.method_defined?(:dependents) ? superclass.dependents.dup : []
```
Wouldn't this then modify superclass' dependents if both super and subclass define dependents?

saghm replied:
I don't think so? The actual method call is `superclass.dependents.dup`, so it isn't reusing the reference to the superclass's array.

saghm replied:
Good point! Let me know what you think of the changes I just pushed

saghm commented:
Just pushed the changes we made in the pair-coding session

p-mongo commented on lib/mongoid/association/depending.rb:
```diff
@@ -10,7 +10,15 @@ module Depending
 
       included do
         class_attribute :dependents
+        class_attribute :dependents_owner
```
Can this be marked `@api private` please.

p-mongo commented on lib/mongoid/association/depending.rb:
```diff
@@ -10,7 +10,15 @@ module Depending
+      end
+
+      class_methods do
+        def _all_dependents
```
And this also `@api private`.

p-mongo commented on lib/mongoid/association/depending.rb:
```diff
@@ -10,7 +10,15 @@ module Depending
+
+      class_methods do
+        def _all_dependents
+          dependents + (superclass.respond_to?(:_all_dependents) ? superclass._all_dependents : [])
```
This needs to deduplicate the list by association name.

saghm replied:
Oh, good catch!

saghm commented:
Suggested changes have been made
